packages/components/src/higher-order/with-filters/index.js

```diff
@@ -24,45 +24,82 @@ const ANIMATION_FRAME_PERIOD = 16;
  */
 export default function withFilters( hookName ) {
 	return createHigherOrderComponent( ( OriginalComponent ) => {
-		return class FilteredComponent extends Component {
-			/** @inheritdoc */
+		const namespace = 'core/with-filters-' + hookName;
+
+		/**
+		 * Since filtering is applied to the component, each filtered instance
+		 * can reuse a shared reference to the definition. This optimizes to
+		 * avoid excessive calls to `applyFilters` when many instances exist.
+		 *
+		 * @type {Component}
+		 */
+		let FilteredComponent = applyFilters( hookName, OriginalComponent );
+
+		class FilteredComponentRenderer extends Component {
 			constructor( props ) {
 				super( props );
 
-				this.onHooksUpdated = this.onHooksUpdated.bind( this );
-				this.Component = applyFilters( hookName, OriginalComponent );
-				this.namespace = uniqueId( 'core/with-filters/component-' );
-				this.throttledForceUpdate = debounce( () => {
-					this.Component = applyFilters( hookName, OriginalComponent );
-					this.forceUpdate();
-				}, ANIMATION_FRAME_PERIOD );
+				this.throttledForceUpdate = debounce(
```

Comment:
We could probably go one step further and move the debounce to a single shared function, calling each instance's `forceUpdate` directly. This has a couple advantages:

- The constructor can be dropped altogether
- Avoids need to cancel in `componentWillUnmount`
- Automatically accounts for unmounting of components, since the callback would only evaluate `FilteredComponentRenderer.instances` after the debounce delay

Since it requires a bit more reorganization, I'm a bit inclined to leave this for a subsequent pull request.